BUG: incorrect handling of scipy.sparse.dok formats

keitakurita opened the pull request:
- [x] closes #16179 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``

kristopheryahoo committed:
BUG: incorrect handling of scipy.sparse.dok formats (#16197)

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -191,11 +191,13 @@ def _init_spmatrix(self, data, index, columns, dtype=None,
         for col, rowvals in values.groupby(data.col):
             # get_blocks expects int32 row indices in sorted order
             rows = rowvals.index.values.astype(np.int32)
+            vals = np.array([y for x, y in sorted(rowvals.iteritems())],
```
this is not right, it needs to be in the order of the index.

keitakurita replied:
Which index do you mean? This code should sort the values by the index of rowvals, or am I mistaken?
Sorry, I'm new to the codebase, so I might be misunderstanding something.

jreback replied:
you are passing index below to create

kernc replied:
I believe this to be the whole of a correct fix:
```patch
diff --git a/pandas/core/sparse/frame.py b/pandas/core/sparse/frame.py
index 05c97fa..19186ef 100644
--- a/pandas/core/sparse/frame.py
+++ b/pandas/core/sparse/frame.py
@@ -190,8 +190,8 @@ class SparseDataFrame(DataFrame):
         values = Series(data.data, index=data.row, copy=False)
         for col, rowvals in values.groupby(data.col):
             # get_blocks expects int32 row indices in sorted order
+            rowvals.sort_index(inplace=True)
             rows = rowvals.index.values.astype(np.int32)
-            rows.sort()
             blocs, blens = get_blocks(rows)
 
             sdict[columns[col]] = SparseSeries(
```

keitakurita replied:
@kernc 
Thanks, this is indeed a faster and more elegant solution. I've committed this solution to this branch.
Just out of curiosity though, this does not seem to be fundamentally different behavior from my original solution, since they both seem to be sorting on the same index of rowvals. Am I mistaken?

kernc replied:
You're exactly right; just somewhat more elegant.

keitakurita replied:
@jreback 
Is this solution still problematic? I think sorting the values by their index ensures that the values are in the correct order that aligns them with the index being passed below.

jreback commented on pandas/core/sparse/frame.py:
```diff
@@ -190,8 +190,8 @@ def _init_spmatrix(self, data, index, columns, dtype=None,
         values = Series(data.data, index=data.row, copy=False)
         for col, rowvals in values.groupby(data.col):
             # get_blocks expects int32 row indices in sorted order
+            rowvals.sort_index(inplace=True)
```
``rowvals = rowvals.sort_index()`` is idiomatic

keitakurita replied:
fixed!

jreback commented:
I add the whatsnew for 0.20.1. pls rebase and add a note in bug fixes.

keitakurita commented:
@jreback 
rebased and notes added

jreback commented on doc/source/whatsnew/v0.20.1.txt:
```diff
@@ -66,8 +66,7 @@ Groupby/Resample/Rolling
 
-
-
+- Bug in construction of SparseDataFrame from ``scipy.sparse.dok_matrix`` (:issue:`16179`)
```
can you move to 0.20.2

jreback commented:
can you add the test from the issue as well (with a comment as to the issue number). ping on green.

keitakurita commented:
@jreback 
I've modified the test for the construction of SparseDataFrame from a sparse matrix to actually catch situations where the direct initialization from the array returns a different result. The reason these tests were succeeding in the past was because the array being tested on just happened to return the correct result by chance.
Should I add another, different test for this issue? Or is commenting on this issue where I have changed the tests enough?

jreback commented:
your change looks fine. I would still like another test that is the exact replica from the issue as well. otherwise lgtm. ping on green.

keitakurita commented:
@jreback 
Thanks, understood and tests have been added

jreback commented:
thanks!
